xml_handlers/OpenCoverTar_handler.php

```diff
@@ -0,0 +1,275 @@
+        foreach ($iterator as $fileinfo) {
+            // need the longest extension, so getExtension() won't do here.
+            $ext = substr(strstr($fileinfo->getFilename(), '.'), 1);
+            if ($ext === 'xml') {
+                $this->ParseOpenCoverFile($this->Build->Id, $fileinfo);
+            }
+
+        }
+        // Record parsed coverage info to the database.
+        foreach ($this->CoverageFileLogs as $path => $coverageFileLog) {
+            $coverage = $this->Coverages[$path];
+            $coverageFile = $this->CoverageFiles[$path];
+
+            // Tally up how many lines of code were covered & uncovered.
+            foreach ($coverageFileLog->Lines as $line) {
+                $coverage->Covered = 1;
+                if ($line == 0) {
+                    $coverage->LocUntested += 1;
+                } else {
+                    $coverage->LocTested += 1;
+                }
+            }
+
+            // Save these models to the database.
+            $coverageFile->TrimLastNewline();
+            $coverageFile->Update($this->Build->Id);
+            $coverageFileLog->BuildId = $this->Build->Id;
+            $coverageFileLog->FileId = $coverageFile->Id;
+            $coverageFileLog->Insert();
+
+            // Add this Coverage to our summary.
+            $coverage->CoverageFile = $coverageFile;
+            $coverageSummary->AddCoverage($coverage);
+        }
+
+        // Insert coverage summaries
+        $completedSummaries = array();
+        foreach ($this->CoverageSummaries as $coverageSummary) {
+            if (in_array($coverageSummary->BuildId, $completedSummaries)) {
+                continue;
+            }
+
+            $coverageSummary->Insert();
+            $coverageSummary->ComputeDifference();
+
+            $completedSummaries[] = $coverageSummary->BuildId;
+        }
+
+        // Delete the directory when we're done.
+        DeleteDirectory($dirName);
+        return true;
+    }
+
+
+    /**
+     * Read in the source for each .cs file
+     **/
+    public function readSourceFile($buildid, $fileinfo) {
+       // If the name starts with "TemporaryGenerated", ignore the file
+       if (preg_match("/^TemporaryGenerated/",$fileinfo->getFilename())) {
+           return true;
+       }
+       $path = $fileinfo->getPath() . DIRECTORY_SEPARATOR . $fileinfo->getFilename();
+       $path = str_replace($this->tarDir.'/','',$path);
+       $fileContents = file($fileinfo->getPath() . DIRECTORY_SEPARATOR . $fileinfo->getFilename());
+       $this->getCoverageObjects($path);
+       $inlongComment= false;
+       if ($this->coverageFile) {
+           foreach($fileContents as $key=>$line) {
+              // Matches the beginning of a comment block
+              if(preg_match("/\/[*]+/", trim($line))) {
+                  $inlongComment=true;
+              }
+
+              $this->coverageFile->File .= rtrim($line).'<br>';
+              if((preg_match("/^\/\//", trim($line))) or
```

Comment:
As a matter of preference this logic should probably be flipped so we don't have empty code blocks. 

Also is regex parsing how this is actually done, can this be generated in a more easily parse-able format? If we're trying to glean information about PHP code, is this something we could use an AST parser, or even PHPs builtin reflection API for?

Nitpick: you should just `$line = trim($line)` at the top of all of these.